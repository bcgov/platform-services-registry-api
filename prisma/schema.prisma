// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "mongodb"
  url      = "mongodb+srv://challenge:uYuMturLtSasyUig@cluster0.v8ikp.mongodb.net/v2?retryWrites=true&w=majority"
}

model User {
  id                                        String                @id @default(auto()) @map("_id") @db.ObjectId
  firstName                                 String
  lastName                                  String
  email                                     String                @unique
  githubId                                  String
  ministry                                  String
  archived                                  Boolean               @default(false)
  created                                   DateTime              @default(now())
  lastSeen                                  DateTime              @updatedAt
  privateCloudProjectOwner                  PrivateCloudProject[] @relation("projectOwner")
  privateCloudProjectPrimaryTechnicalLead   PrivateCloudProject[] @relation("primaryTechnicalLead")
  privateCloudProjectSecondaryTechnicalLead PrivateCloudProject[] @relation("secondaryTechnicalLead")
  privateCloudRequestProjectOwner           PrivateCloudRequest[] @relation("requestProjectOwner")
  privateCloudRequestPrimaryTechnicalLead   PrivateCloudRequest[] @relation("requestPrimaryTechnicalLead")
  privateCloudRequestSecondaryTechnicalLead PrivateCloudRequest[] @relation("requestSecondaryTechnicalLead")
}

model PrivateCloudRequest {
  id                       String                     @id @default(auto()) @map("_id") @db.ObjectId
  createdByEmail           String
  decisionMakerEmail       String?
  projectOwnerId           String                     @db.ObjectId
  projectOwner             User                       @relation("requestProjectOwner", fields: [projectOwnerId], references: [id])
  primaryTechnicalLeadId   String                     @db.ObjectId
  primaryTechnicalLead     User                       @relation("requestPrimaryTechnicalLead", fields: [primaryTechnicalLeadId], references: [id])
  secondaryTechnicalLeadId String?                    @db.ObjectId
  secondaryTechnicalLead   User?                      @relation("requestSecondaryTechnicalLead", fields: [secondaryTechnicalLeadId], references: [id])
  type                     String
  status                   String
  active                   Boolean                    @default(true)
  created                  DateTime                   @default(now())
  decisionDate             DateTime?
  projectId                String?                    @unique @db.ObjectId
  project                  PrivateCloudProject?       @relation("requestProject", fields: [projectId], references: [id])
  requestedProject         PrivateCloudProjectEmbeded
}

model PrivateCloudProject {
  id                       String               @id @default(auto()) @map("_id") @db.ObjectId
  name                     String
  licencePlate             String
  archived                 Boolean
  created                  DateTime             @default(now())
  createdByEmail           String
  description              String
  status                   String
  projectOwnerId           String               @db.ObjectId
  projectOwner             User                 @relation("projectOwner", fields: [projectOwnerId], references: [id])
  primaryTechnicalLeadId   String               @db.ObjectId
  primaryTechnicalLead     User                 @relation("primaryTechnicalLead", fields: [primaryTechnicalLeadId], references: [id])
  secondaryTechnicalLeadId String               @db.ObjectId
  secondaryTechnicalLead   User                 @relation("secondaryTechnicalLead", fields: [secondaryTechnicalLeadId], references: [id])
  ministry                 String
  cluster                  String
  productionQuota          Quota
  testQuota                Quota
  developmentQuota         Quota
  toolsQuota               Quota
  commonComponents         CommonComponents
  request                  PrivateCloudRequest? @relation("requestProject")
}

model ArchivedPrivateCloudRequest {
  id                       String                     @id @default(auto()) @map("_id") @db.ObjectId
  createdById              String                     @unique @db.ObjectId
  decisionMakerId          String                     @unique @db.ObjectId
  projectOwnerId           String                     @db.ObjectId
  primaryTechnicalLeadId   String                     @db.ObjectId
  secondaryTechnicalLeadId String                     @db.ObjectId
  type                     String
  status                   String
  active                   Boolean
  created                  DateTime                   @default(now())
  decisionDate             DateTime
  projectId                String
  requestedProject         PrivateCloudProjectEmbeded
}

type PrivateCloudProjectEmbeded {
  name                        String
  licencePlate                String
  archived                    Boolean          @default(false)
  created                     DateTime         @default(now())
  createdByEmail              String
  description                 String
  projectOwnerEmail           String
  primaryTechnicalLeadEmail   String
  secondaryTechnicalLeadEmail String
  ministry                    String
  cluster                     String
  productionQuota             Quota
  testQuota                   Quota
  developmentQuota            Quota
  toolsQuota                  Quota
  commonComponents            CommonComponents
}

type Quota {
  cpuRequests     Float @default(0.5)
  cpuLimits       Float @default(1.5)
  memoryRequests  Int   @default(2)
  memoryLimits    Int   @default(4)
  storageBlock    Int   @default(1)
  storageFile     Int   @default(1)
  storageBackup   Int   @default(512)
  storageCapacity Int   @default(1)
  storagePvcCount Int   @default(5)
  snapshotCount   Int   @default(5)
}

type CommonComponents {
  addressAndGeolocation              String?
  workflowManagement                 String?
  formDesignAndSubmission            String?
  identityManagement                 String?
  paymentServices                    String?
  documentManagement                 String?
  endUserNotificationAndSubscription String?
  publishing                         String?
  businessIntelligence               String?
  other                              String?
}
