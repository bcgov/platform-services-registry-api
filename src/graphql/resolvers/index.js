import { GraphQLDateTime } from "graphql-scalars";
import { users, user, usersByIds, me } from "./Query/Users";
import {
  privateCloudProjects,
  privateCloudProject,
  publicCloudProjects,
  publicCloudProject,
  projects,
} from "./Query/Projects";
import {
  privateCloudRequests,
  privateCloudRequest,
  publicCloudRequests,
  publicCloudRequest,
  requests,
} from "./Query/Requests";
import createPrivateCloudProjectRequest from "./Mutation/CreatePrivateCloudProjectRequest";
import createUser from "./Mutation/CreateUser";
import signUp from "./Mutation/SignUp";
import createPrivateCloudProjectEditRequest from "./Mutation/CreatePrivateCloudProjectEditRequest";
import makePrivateCloudRequestDecision from "./Mutation/MakePrivateCloudRequestDecision";
import PrivateCloudProject from "./PrivateCloudProject";
import User from "./User";
import Project from "./Project";
import Request from "./Request";
import Ministry from "./enum/Ministry";
import Environment from "./enum/Environment";
import RequestStatus from "./enum/RequestStatus";
import ProjectStatus from "./enum/ProjectStatus";
import RequestType from "./enum/RequestType";
import PublicCloudPlatform from "./enum/PublicCloudPlatform";
import Platform from "./enum/Platform";
import RequestDecision from "./enum/RequestDecision";

const resolvers = {
  Query: {
    users,
    user,
    usersByIds,
    me,
    privateCloudProjects,
    privateCloudProject,
    publicCloudProjects,
    publicCloudProject,
    projects,
    privateCloudRequests,
    privateCloudRequest,
    publicCloudRequests,
    publicCloudRequest,
    requests,
  },
  Mutation: {
    signUp,
    createUser,
    createPrivateCloudProjectRequest,
    createPrivateCloudProjectEditRequest,
    makePrivateCloudRequestDecision
  },
  PrivateCloudProject,
  User,
  DateTime: GraphQLDateTime,
  ProjectStatus,
  Project,
  Request,
  RequestStatus,
  RequestType,
  Ministry,
  Environment,
  PublicCloudPlatform,
  Platform,
  RequestDecision,
};

export default resolvers;
