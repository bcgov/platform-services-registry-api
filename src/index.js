require("dotenv").config();
import { ApolloServer } from "apollo-server-express";
import { ApolloServerPluginDrainHttpServer } from "apollo-server-core";
import { makeExecutableSchema } from "@graphql-tools/schema";
import { configureKeycloak } from "./auth/config";
import { KeycloakContext, KeycloakTypeDefs } from "keycloak-connect-graphql";
import { applyDirectiveTransformers } from "./auth/transformers";
import express from "express";
import http from "http";
import typeDefs from "./graphql/typeDefs";
import resolvers from "./graphql/resolvers";
import { MongoClient } from "mongodb";
import MongoHelpers from "./dataSources/MongoHelpers";

async function startApolloServer(typeDefs, resolvers) {
  const client = new MongoClient(process.env.MONGO_URI);
  client.connect();

  let schema = makeExecutableSchema({
    typeDefs: [KeycloakTypeDefs, typeDefs],
    resolvers,
  });

  schema = applyDirectiveTransformers(schema);

  const app = express();

  const graphqlPath = "/graphql";
  const { keycloak } = configureKeycloak(app, graphqlPath);

  const httpServer = http.createServer(app);

  const server = new ApolloServer({
    schema,
    resolvers,
    dataSources: () => ({
      users: new MongoHelpers(client.db().collection("users")),
      privateCloudProjects: new MongoHelpers(
        client.db().collection("privateCloudProjects")
      ),
      publicCloudProjects: new MongoHelpers(
        client.db().collection("publicCloudProjects")
      ),
      privateCloudRequests: new MongoHelpers(
        client.db().collection("privateCloudRequests")
      ),
      publicCloudRequests: new MongoHelpers(
        client.db().collection("publicCloudRequests")
      ),
      privateCloudRequestedProjects: new MongoHelpers(
        client.db().collection("privateCloudRequestedProjects")
      ),
      publicCloudRequestedProjects: new MongoHelpers(
        client.db().collection("publicCloudRequestedProjects")
      ),
    }),
    context: ({ req }) => {
      return ({kauth: new KeycloakContext({ req }, keycloak)})
    },
    plugins: [ApolloServerPluginDrainHttpServer({ httpServer })],
  });

  await server.start();
  server.applyMiddleware({ app });
  await new Promise((resolve) => httpServer.listen({ port: 4000 }, resolve));
  console.log(`ðŸš€ Server ready at http://localhost:4000${server.graphqlPath}`);
}

startApolloServer(typeDefs, resolvers);
