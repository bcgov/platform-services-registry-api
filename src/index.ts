import './env.js';
import { ApolloServer } from '@apollo/server';
import { expressMiddleware } from '@apollo/server/express4';
import { ApolloServerPluginDrainHttpServer } from '@apollo/server/plugin/drainHttpServer';
import { makeExecutableSchema } from '@graphql-tools/schema';
import { KeycloakContext, KeycloakTypeDefs } from 'keycloak-connect-graphql';
import express from 'express';
import http from 'http';
import cors from 'cors';
import bodyParser from 'body-parser';
import resolvers from './resolvers/index.js';
import { readFileSync } from 'fs';
import configureKeycloak from './auth/config.js';
import { DIRECTIVES } from '@graphql-codegen/typescript-mongodb';
import applyDirectiveTransformers from './transformers/index.js';
import { PrismaClient } from '@prisma/client';
import {
  privateCloudProvisionerCallbackHandler,
  publicCloudProvisionerCallbackHandler,
  getReProvisionNatsMessage,
  getIdsForCluster,
  getDatabaseHealthCheck,
} from './controllers/index.js';
import chesService from './ches/index.js';
import { connectToDatabase } from './db.js';

const typeDefs = readFileSync('./schema.graphql', { encoding: 'utf-8' });

export interface ContextValue {
  kauth: KeycloakContext;
  prisma: PrismaClient;
  authRoles: string[];
  authEmail: string;
  chesService: typeof chesService;
}

export const prisma = new PrismaClient();

let schema = makeExecutableSchema({
  typeDefs: [KeycloakTypeDefs, typeDefs, DIRECTIVES],
  resolvers,
});

schema = applyDirectiveTransformers(schema);

export const app = express();
const { keycloak } = configureKeycloak(app, '/');
const httpServer = http.createServer(app);

export const server = new ApolloServer<ContextValue>({
  schema,
  plugins: [ApolloServerPluginDrainHttpServer({ httpServer })],
  introspection: true,
});

await server.start();

await connectToDatabase();

app.use(
  '/graphql',
  cors<cors.CorsRequest>(),
  bodyParser.json(),
  expressMiddleware(server, {
    context: async ({ req }) => {
      // @ts-ignore
      const kauth = new KeycloakContext({ req }, keycloak);
      // @ts-ignore
      const email = kauth?.accessToken?.content?.email;
      // @ts-ignore
      const resource_access = kauth?.accessToken?.content?.resource_access;
      const { roles } = resource_access?.[process.env.AUTH_RESOURCE] || {
        roles: [],
      };

      return {
        kauth: kauth,
        prisma,
        authRoles: roles,
        authEmail: email,
        chesService,
      };
    },
  })
);

// The below code is important for auth to work
app.use(express.json());
app.use(express.urlencoded({ extended: true }));

app.get(
  '/api/v1/provision/sync/:profile_id/provisioned-profile-bot-json',
  keycloak.protect(),
  getReProvisionNatsMessage
);

app.get(
  '/api/v1/provision/sync/provisioned-profile-ids',
  keycloak.protect(),
  getIdsForCluster
);

app.get('/api/v1/database-health-check', getDatabaseHealthCheck);

// app.post("/namespace", keycloak.protect(), provisionerCallbackHandler);
app.post('/namespace', privateCloudProvisionerCallbackHandler);
app.post('/public-cloud', publicCloudProvisionerCallbackHandler);

// app.post("/predeletion-check")

await new Promise<void>((resolve) =>
  httpServer.listen({ port: 4000 }, resolve)
);
console.log(`ðŸš€ Server ready at http://localhost:4000/graphql`);
