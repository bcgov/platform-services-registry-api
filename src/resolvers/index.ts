import {
  QueryResolvers,
  MutationResolvers,
  Resolvers
} from "../__generated__/resolvers-types";
import * as userMutations from "./mutations/users.js";
import * as requestMutations from "./mutations/requests.js";
import * as userQueries from "./queries/users.js";
import User from "./user.js";
import * as enums from "./enum.js";

const mutations: MutationResolvers = {
  Mutation: {
    ...userMutations,
    ...requestMutations
  }
};

const queries: QueryResolvers = {
  Query: { ...userQueries }
};

const resolvers: Resolvers = { ...mutations, ...queries, ...enums, User };

export default resolvers;
