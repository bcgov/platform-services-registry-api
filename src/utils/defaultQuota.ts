const defaultQuota = {
  cpuRequests: 0.5,
  cpuLimits: 1.5,
  memoryRequests: 2,
  memoryLimits: 4,
  storageBlock: 1,
  storageFile: 1,
  storageBackup: 512,
  storageCapacity: 1,
  storagePvcCount: 60,
  snapshotCount: 5
};

export default defaultQuota;