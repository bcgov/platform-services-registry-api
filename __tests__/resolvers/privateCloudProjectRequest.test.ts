import "../../src/env";
import {
  Cluster,
  Ministry,
  CommonComponentsInput,
  CreateUserInput,
  RequestDecision,
  RequestType,
  CommonComponentsOptions,
  DefaultCpuOptions,
  DefaultMemoryOptions,
  DefaultStorageOptions,
} from "../../src/__generated__/resolvers-types.js";
import {
  DefaultCpuOptions as DefaultCpuOptionsEnum,
  DefaultMemoryOptions as DefaultMemoryOptionsEnum,
  DefaultStorageOptions as DefaultStorageOptionsEnum,
} from "../../src/resolvers/enum";
import resolvers from "../../src/resolvers/index.js";
import { ApolloServer } from "@apollo/server";
import { makeExecutableSchema } from "@graphql-tools/schema";
import { KeycloakContext, KeycloakTypeDefs } from "keycloak-connect-graphql";
import { readFileSync } from "fs";
import { DIRECTIVES } from "@graphql-codegen/typescript-mongodb";
import applyDirectiveTransformers from "../../src/transformers/index.js";
import { PrismaClient } from "@prisma/client";
import req from "../../src/auth/kauthContextMock.js";
import assert from "assert";
import supertest from "supertest";
import {
  mockProjectOwner,
  mockPrimaryTechnicalLead,
  mockSecondaryTechnicalLead,
  mockProjectA,
  mockProjectB,
} from "../../__mocks__/constants.js";
import { defaultQuota } from "../../src/utils/defaultQuota";

interface ContextValue {
  kauth: KeycloakContext;
  prisma: PrismaClient;
  authEmail: string;
  authRoles: string[];
}

const prisma = new PrismaClient();

const contextValue: ContextValue = {
  prisma: prisma,
  // @ts-ignore
  kauth: new KeycloakContext({ req }),
  authEmail: "oamar.kanji@gov.bc.ca",
  authRoles: ["admin"],
};

const typeDefs = readFileSync("./schema.graphql", { encoding: "utf-8" });

let schema = makeExecutableSchema({
  typeDefs: [KeycloakTypeDefs, typeDefs, DIRECTIVES],
  resolvers,
});

schema = applyDirectiveTransformers(schema);

const server = new ApolloServer<ContextValue>({
  schema,
});

let createRequestIdA;
let createRequestIdB;

describe("Request tests", () => {
  beforeAll(async () => {
    await prisma.user.createMany({
      data: [
        mockProjectOwner,
        mockPrimaryTechnicalLead,
        mockSecondaryTechnicalLead,
      ],
    });

    await prisma.privateCloudProject.create({
      data: mockProjectA,
    });

    await prisma.privateCloudProject.create({
      data: mockProjectB,
    });
  });

  afterAll(async () => {
    const deleteRequests = prisma.privateCloudRequest.deleteMany();
    const deleteProjects = prisma.privateCloudProject.deleteMany();
    const deleteRequestedProjects =
      prisma.privateCloudRequestedProject.deleteMany();
    const deleteUsers = prisma.user.deleteMany();

    await prisma.$transaction([
      deleteRequests,
      deleteProjects,
      deleteRequestedProjects,
      deleteUsers,
    ]);

    await prisma.$disconnect();
  });

  test("creates a private cloud project request without secondary technical lead", async () => {
    const CREATE_PRIVATE_CLOUD_PROJECT_REQUEST = `mutation PrivateCloudProjectRequest(
        $name: String!,
        $description: String!,
        $ministry: Ministry!,
        $cluster: Cluster!,
        $commonComponents: CommonComponentsInput!,
        $projectOwner: CreateUserInput!,
        $primaryTechnicalLead: CreateUserInput!,
        $secondaryTechnicalLead: CreateUserInput
      ) {
        privateCloudProjectRequest(
          name: $name,
          description: $description,
          ministry: $ministry,
          cluster: $cluster,
          commonComponents: $commonComponents,
          projectOwner: $projectOwner,
          primaryTechnicalLead: $primaryTechnicalLead,
          secondaryTechnicalLead: $secondaryTechnicalLead
        ) {
          id
          active
          createdBy {
            email
          }
          decisionDate
          decisionMaker {
            email
          }
          decisionStatus
          project {
            name
          }
          requestedProject {
            id
          }
      }
    }`;

    const name: string = "Test Project";
    const description: string = "Test Description";
    const ministry: Ministry = Ministry.Agri;
    const cluster: Cluster = Cluster.Klab;

    const commonComponents: CommonComponentsInput = {
      other: "test",
      noServices: false,
    };

    const projectOwner: CreateUserInput = {
      email: "oamar.kanji@gov.bc.ca",
      firstName: "Oamar",
      githubId: "okanji",
      lastName: "Kanji",
      ministry: Ministry.Agri,
    };

    const primaryTechnicalLead: CreateUserInput = {
      email: "xyz@test.com",
      firstName: "Jane",
      githubId: "test123",
      lastName: "Smith",
      ministry: Ministry.Psa,
    };

    const response = await server.executeOperation(
      {
        query: CREATE_PRIVATE_CLOUD_PROJECT_REQUEST,
        variables: {
          name,
          description,
          ministry,
          cluster,
          commonComponents,
          projectOwner,
          primaryTechnicalLead,
        },
      },
      {
        contextValue,
      }
    );

    assert(response.body.kind === "single");
    expect(response.body.singleResult.errors).toBeUndefined();
    expect(response).toMatchSnapshot({
      body: {
        singleResult: {
          data: {
            privateCloudProjectRequest: {
              id: expect.any(String),
              requestedProject: {
                id: expect.any(String),
              },
            },
          },
        },
      },
    });

    // prettier-ignore
    // @ts-ignore
    createRequestIdA = response.body.singleResult.data.privateCloudProjectRequest.id;

    const requests = await prisma.privateCloudRequest.findMany();
    expect(requests.length).toBe(1);

    const requestedProject =
      await prisma.privateCloudRequestedProject.findUnique({
        where: {
          // prettier-ignore
          // @ts-ignore
          id: response.body.singleResult.data.privateCloudProjectRequest.requestedProject.id,
        },
      });

    expect(requestedProject).not.toBeNull();
    expect(requestedProject?.name).toBe(name);
    expect(requestedProject?.description).toBe(description);
    expect(requestedProject?.ministry).toBe(ministry);
    expect(requestedProject?.cluster).toBe(cluster);

    const projectOwnerUser = await prisma.user.findUnique({
      where: {
        email: projectOwner.email,
      },
    });

    expect(projectOwnerUser).not.toBeNull();
    expect(projectOwnerUser?.firstName).toBe(projectOwner.firstName);
    expect(projectOwnerUser?.lastName).toBe(projectOwner.lastName);
    expect(projectOwnerUser?.githubId).toBe(projectOwner.githubId);
    expect(projectOwnerUser?.ministry).toBe(projectOwner.ministry);

    const primaryTechnicalLeadUser = await prisma.user.findUnique({
      where: {
        email: primaryTechnicalLead.email,
      },
    });

    expect(primaryTechnicalLeadUser).not.toBeNull();
    expect(primaryTechnicalLeadUser?.firstName).toBe(
      primaryTechnicalLead.firstName
    );
    expect(primaryTechnicalLeadUser?.lastName).toBe(
      primaryTechnicalLead.lastName
    );
    expect(primaryTechnicalLeadUser?.githubId).toBe(
      primaryTechnicalLead.githubId
    );
    expect(primaryTechnicalLeadUser?.ministry).toBe(
      primaryTechnicalLead.ministry
    );
  });

  test("Creates a private cloud project request with secondary technical lead", async () => {
    const CREATE_PRIVATE_CLOUD_PROJECT_REQUEST = `
    mutation PrivateCloudProjectRequest(
      $name: String!,
      $description: String!,
      $ministry: Ministry!,
      $cluster: Cluster!,
      $commonComponents: CommonComponentsInput!,
      $projectOwner: CreateUserInput!,
      $primaryTechnicalLead: CreateUserInput!,
      $secondaryTechnicalLead: CreateUserInput
    ) {
      privateCloudProjectRequest(
        name: $name,
        description: $description,
        ministry: $ministry,
        cluster: $cluster,
        commonComponents: $commonComponents,
        projectOwner: $projectOwner,
        primaryTechnicalLead: $primaryTechnicalLead,
        secondaryTechnicalLead: $secondaryTechnicalLead
      ) {
        id
        active
        createdBy {
          email
        }
        decisionDate
        decisionMaker {
          email
        }
        decisionStatus
        project {
          name
        }
    }
  }`;

    const name: string = "Test Project";
    const description: string = "Test Description";
    const ministry: Ministry = Ministry.Agri;
    const cluster: Cluster = Cluster.Klab;

    const commonComponents: CommonComponentsInput = {
      other: "test",
      noServices: false,
    };

    const projectOwner: CreateUserInput = {
      email: "oamar.kanji@gov.bc.ca",
      firstName: "Oamar",
      githubId: "okanji",
      lastName: "Kanji",
      ministry: Ministry.Agri,
    };

    const primaryTechnicalLead: CreateUserInput = {
      email: "xyz@test.com",
      firstName: "Jane",
      githubId: "test123",
      lastName: "Smith",
      ministry: Ministry.Psa,
    };

    const secondaryTechnicalLead: CreateUserInput = {
      email: "testSecondaryTechnicalLead@test.com",
      firstName: "John",
      githubId: "test456",
      lastName: "Doe",
      ministry: Ministry.Psa,
    };

    const response = await server.executeOperation(
      {
        query: CREATE_PRIVATE_CLOUD_PROJECT_REQUEST,
        variables: {
          name,
          description,
          ministry,
          cluster,
          commonComponents,
          projectOwner,
          primaryTechnicalLead,
          secondaryTechnicalLead,
        },
      },
      {
        contextValue,
      }
    );

    assert(response.body.kind === "single");
    expect(response.body.singleResult.errors).toBeUndefined();
    expect(response).toMatchSnapshot({
      body: {
        singleResult: {
          data: {
            privateCloudProjectRequest: {
              id: expect.any(String),
            },
          },
        },
      },
    });

    const requests = await prisma.privateCloudRequest.findMany();
    expect(requests.length).toBe(2);

    const secondaryTechnicalLeadUser = await prisma.user.findUnique({
      where: {
        email: secondaryTechnicalLead.email,
      },
    });

    // @ts-ignore
    // prettier-ignore
    createRequestIdB = response.body.singleResult.data.privateCloudProjectRequest.id;

    expect(secondaryTechnicalLeadUser).not.toBeNull();
    expect(secondaryTechnicalLeadUser?.firstName).toBe(
      secondaryTechnicalLead.firstName
    );
    expect(secondaryTechnicalLeadUser?.lastName).toBe(
      secondaryTechnicalLead.lastName
    );
    expect(secondaryTechnicalLeadUser?.githubId).toBe(
      secondaryTechnicalLead.githubId
    );
    expect(secondaryTechnicalLeadUser?.ministry).toBe(
      secondaryTechnicalLead.ministry
    );
  });

  test("Approve a private cloud project request", async () => {
    const MAKE_PRIVATE_CLOUD_PROJECT_REQUEST_DECISION = `
    mutation PrivateCloudRequestDecision(
      $requestId: ID!,
      $decision: RequestDecision!) {
      privateCloudRequestDecision(requestId: $requestId, decision: $decision) {
        id
        decisionMaker {
          email
        }
        active
        createdBy {
          email
        }
        type
        decisionStatus
      }
    }`;

    const response = await server.executeOperation(
      {
        query: MAKE_PRIVATE_CLOUD_PROJECT_REQUEST_DECISION,
        variables: {
          decision: RequestDecision.Approved,
          requestId: createRequestIdA,
        },
      },
      {
        contextValue,
      }
    );

    const request = await prisma.privateCloudRequest.findUnique({
      where: {
        id: createRequestIdA,
      },
    });

    console.log(request);

    expect(response).toMatchSnapshot({
      body: {
        singleResult: {
          data: {
            privateCloudRequestDecision: {
              id: expect.any(String),
            },
          },
        },
      },
    });

    // @ts-ignore
    const responseErrors = response?.body?.singleResult?.errors;

    expect(request).not.toBeNull();
    expect(request?.decisionStatus).toBe(RequestDecision.Approved);
    expect(request?.createdByEmail).toBe(contextValue.authEmail);
  });

  test("Reject a private cloud project request", async () => {
    const MAKE_PRIVATE_CLOUD_PROJECT_REQUEST_DECISION = `mutation PrivateCloudRequestDecision(
      $requestId: ID!,
      $decision: RequestDecision!) {
      privateCloudRequestDecision(requestId: $requestId, decision: $decision) {
        id
        decisionMaker {
          email
        }
        active
        createdBy {
          email
        }
        type
        decisionStatus
      }
    }`;

    const response = await server.executeOperation(
      {
        query: MAKE_PRIVATE_CLOUD_PROJECT_REQUEST_DECISION,
        variables: {
          decision: RequestDecision.Rejected,
          requestId: createRequestIdB,
        },
      },
      {
        contextValue,
      }
    );

    const request = await prisma.privateCloudRequest.findUnique({
      where: {
        id: createRequestIdB,
      },
    });

    expect(response).toMatchSnapshot({
      body: {
        singleResult: {
          data: {
            privateCloudRequestDecision: {
              id: expect.any(String),
            },
          },
        },
      },
    });

    expect(request).not.toBeNull();
    expect(request?.decisionStatus).toBe(RequestDecision.Rejected);
    expect(request?.createdByEmail).toBe(contextValue.authEmail);
  });

  test("Delete request", async () => {
    const DELETE_PRIVATE_CLOUD_PROJECT_REQUEST = `mutation PrivateCloudProjectDeleteRequest($projectId: ID!) {
      privateCloudProjectDeleteRequest(projectId: $projectId) {
        id
        active
        createdBy {
          email
        }
        decisionDate
        decisionMaker {
          email
        }
        type
        decisionStatus
        project {
          name
        }
      }
    }`;

    const response = await server.executeOperation(
      {
        query: DELETE_PRIVATE_CLOUD_PROJECT_REQUEST,
        variables: {
          projectId: mockProjectA.id,
        },
      },
      {
        contextValue,
      }
    );

    const request = await prisma.privateCloudRequest.findUnique({
      where: {
        // @ts-ignore
        id: response.body.singleResult.data.privateCloudProjectDeleteRequest.id,
      },
    });

    expect(response).toMatchSnapshot({
      body: {
        singleResult: {
          data: {
            privateCloudProjectDeleteRequest: {
              id: expect.any(String),
            },
          },
        },
      },
    });

    expect(request).not.toBeNull();
    expect(request?.active).toBe(true);
    expect(request?.type).toBe(RequestType.Delete);
  });

  test("Makes an edit request", async () => {
    const EDIT_REQUEST = `mutation Mutation(
      $projectId: ID!,
      $name: String!,
      $description: String!,
      $ministry: Ministry!,
      $commonComponents: CommonComponentsInput!,
      $projectOwner: CreateUserInput!,
      $primaryTechnicalLead: CreateUserInput!,
      $secondaryTechnicalLead: CreateUserInput
      $testQuota: QuotaInput!,
      $productionQuota: QuotaInput!,
      $developmentQuota: QuotaInput!,
      $toolsQuota: QuotaInput!,
    ) {
      privateCloudProjectEditRequest(
        projectId: $projectId,
        name: $name,
        description: $description,
        ministry: $ministry,
        commonComponents: $commonComponents,
        projectOwner: $projectOwner,
        primaryTechnicalLead: $primaryTechnicalLead,
        secondaryTechnicalLead: $secondaryTechnicalLead
        testQuota: $testQuota,
        productionQuota: $productionQuota,
        developmentQuota: $developmentQuota,
        toolsQuota: $toolsQuota,
      ) {
        id
      }
    }`;

    const variables = {
      projectId: mockProjectB.id,
      name: "new name",
      description: "new description",
      projectOwner: {
        email: "oamarkanji@gmail.com",
        firstName: "testA",
        lastName: "testA",
        githubId: "testA",
        ministry: Ministry.Agri,
      },
      primaryTechnicalLead: {
        email: "new@test.com",
        firstName: "testA",
        lastName: "testA",
        githubId: "testA",
        ministry: Ministry.Agri,
      },
      commonComponents: {
        identityManagement: CommonComponentsOptions.PlanningToUse,
        noServices: false,
      },
      productionQuota: {
        cpu: DefaultCpuOptions.CpuRequest_0_5Limit_1_5,
        memory: DefaultMemoryOptions.MemoryRequest_64Limit_128,
        storage: DefaultStorageOptions.Storage_16,
      },
      testQuota: {
        cpu: DefaultCpuOptions.CpuRequest_0_5Limit_1_5,
        memory: DefaultMemoryOptions.MemoryRequest_64Limit_128,
        storage: DefaultStorageOptions.Storage_16,
      },
      developmentQuota: {
        cpu: DefaultCpuOptions.CpuRequest_0_5Limit_1_5,
        memory: DefaultMemoryOptions.MemoryRequest_64Limit_128,
        storage: DefaultStorageOptions.Storage_16,
      },
      toolsQuota: {
        cpu: DefaultCpuOptions.CpuRequest_0_5Limit_1_5,
        memory: DefaultMemoryOptions.MemoryRequest_64Limit_128,
        storage: DefaultStorageOptions.Storage_16,
      },
      ministry: Ministry.Agri,
    };

    const response = await server.executeOperation(
      {
        query: EDIT_REQUEST,
        variables,
      },
      {
        contextValue,
      }
    );

    expect(response).toMatchSnapshot({
      body: {
        singleResult: {
          data: {
            privateCloudProjectEditRequest: {
              id: expect.any(String),
            },
          },
        },
      },
    });

    const request = await prisma.privateCloudRequest.findUnique({
      where: {
        // @ts-ignore
        id: response.body.singleResult.data.privateCloudProjectEditRequest.id,
      },
      include: {
        project: true,
        requestedProject: true,
      },
    });

    expect(request).not.toBeNull();
    expect(request?.active).toBe(true);
    expect(request?.type).toBe(RequestType.Edit);
    expect(request?.project?.name).toBe(mockProjectB.name);
    expect(request?.requestedProject?.name).toBe(variables.name);
    expect(request?.project?.description).toBe(mockProjectB.description);
    expect(request?.requestedProject?.description).toBe(variables.description);
    expect(request?.project?.cluster).toBe(mockProjectB.cluster);
    expect(request?.requestedProject?.cluster).toBe(mockProjectB.cluster);
    expect(request?.project?.commonComponents).toEqual(
      mockProjectB.commonComponents
    );
    expect(request?.requestedProject?.commonComponents).toEqual({
      addressAndGeolocation: null,
      workflowManagement: null,
      formDesignAndSubmission: null,
      identityManagement: "PLANNING_TO_USE",
      paymentServices: null,
      documentManagement: null,
      endUserNotificationAndSubscription: null,
      publishing: null,
      businessIntelligence: null,
      other: null,
      noServices: false,
    });
    expect(request?.project?.productionQuota).toEqual(
      mockProjectB.productionQuota
    );
    expect(request?.requestedProject?.productionQuota).toEqual({
      ...defaultQuota,
      ...DefaultCpuOptionsEnum[variables.productionQuota.cpu],
      ...DefaultMemoryOptionsEnum[variables.productionQuota.memory],
      ...DefaultStorageOptionsEnum[variables.productionQuota.storage],
    });
  });
});
