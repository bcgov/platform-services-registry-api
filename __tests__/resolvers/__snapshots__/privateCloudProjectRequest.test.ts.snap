// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Request tests creates a private cloud project request 1`] = `
{
  "body": {
    "kind": "single",
    "singleResult": {
      "data": {
        "privateCloudProjectRequest": {
          "active": true,
          "createdBy": {
            "email": "oamar.kanji@gov.bc.ca",
          },
          "decisionDate": null,
          "decisionMaker": null,
          "decisionStatus": "PENDING",
          "project": null,
        },
      },
      "errors": undefined,
    },
  },
  "http": {
    "headers": Map {
      "cache-control" => "no-store",
    },
    "status": undefined,
  },
}
`;
